pkg/gce-pd-csi-driver/node.go

```diff
@@ -365,7 +368,70 @@ func (ns *GCENodeServer) NodeGetVolumeStats(ctx context.Context, req *csi.NodeGe
 }
 
 func (ns *GCENodeServer) NodeExpandVolume(ctx context.Context, req *csi.NodeExpandVolumeRequest) (*csi.NodeExpandVolumeResponse, error) {
-	return nil, status.Error(codes.Unimplemented, fmt.Sprintf("NodeExpandVolume is not yet implemented"))
+	volumeID := req.GetVolumeId()
+	if len(volumeID) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "ControllerExpandVolume volume ID must be provided")
+	}
+	capacityRange := req.GetCapacityRange()
+	reqBytes, err := getRequestCapacity(capacityRange)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("ControllerExpandVolume capacity range is invalid: %v", err))
+	}
+
+	volumePath := req.GetVolumePath()
+	if len(volumePath) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "ControllerExpandVolume volume path must be provided")
+	}
+
+	volKey, err := common.VolumeIDToKey(volumeID)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("ControllerExpandVolume volume ID is invalid: %v", err))
+	}
+
+	devicePath, err := ns.getDevicePath(volumeID, "")
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("ControllerExpandVolume error when getting device path for %s: %v", volumeID, err))
+	}
+
+	resizer := resizefs.NewResizeFs(ns.Mounter)
+	resized, err := resizer.Resize(devicePath, volumePath)
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("ControllerExpandVolume error when resizing volume %s: %v", volKey.String(), err))
+
+	}
+
+	// Verify the resize actually happened.
+	format, err := ns.Mounter.GetDiskFormat(devicePath)
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("ControllerExpandVolume error checking format for device %s: %v", devicePath, err))
+	}
+	if format == "" {
+		// Block volume
+		gotSizeGb, err := ns.getBlockSizeGb(devicePath)
+		if err != nil {
+			return nil, status.Error(codes.Internal, fmt.Sprintf("ControllerExpandVolume error when getting size of block volume at path %s: %v", devicePath, err))
+		}
+		if gotSizeGb != common.BytesToGb(reqBytes) {
+			return nil, status.Errorf(codes.Internal, "ControllerExpandVolume resize requested for %v but after resize volume was size %v", common.BytesToGb(reqBytes), gotSizeGb)
+		}
+	} else {
+		// Some sort of formatted volume
+		gotSizeGb, err := ns.getFSSizeGb(devicePath)
+
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "ControllerExpandVolume resize could not get fs size of %s: %v", volumePath, err)
+		}
+		if gotSizeGb != common.BytesToGb(reqBytes) {
```

Comment:
exact equality is going to be a problem, also as I mentioned somewhere below formatted disk might be smaller than underlying volume.

Reply:
changed to only check exact equality for the actual block device: https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/pull/317/commits/3a7e29fe2421c66f6d22d73eb6d06e48242f6ad5#diff-37c748634a23b1ed91abeaa8a6d21525R405

Ignoring fs size for now